py_mini_racer/extension/mini_racer_extension.cc

```diff
@@ -395,63 +416,66 @@ BinaryValue *convert_v8_to_binary(ContextInfo *context_info,
-                    return NULL;
+                res->hash_val = static_cast<BinaryValue**>(
+                            malloc(sizeof(*res->hash_val) * hash_len * 2));
+                if (!res->hash_val) {
+                    goto err;
                 }
             } else {
-                res->value = NULL;
+                res->hash_val = NULL;
             }
-            for(uint32_t i=0; i < hash_len; i++) {
-                Local<Value> key = props->Get(i);
-                Local<Value> value = object->Get(key);
+
+            for (uint32_t i = 0; i < hash_len; i++) {
+                Local<Value> pkey = props->Get(i);
+                Local<Value> pvalue = object->Get(pkey);
                 // this may have failed due to Get raising
 
                 if (trycatch.HasCaught()) {
-                    // TODO isolate code that translates execption
-
-                    free(res->value);
-                    free(res);
-                    return NULL;
+                    goto err;
                 }
 
-                BinaryValue *bin_key = convert_v8_to_binary(context_info, key);
-                BinaryValue *bin_value = convert_v8_to_binary(context_info, value);
+                BinaryValue *bin_key = convert_v8_to_binary(context_info, pkey);
+                BinaryValue *bin_value = convert_v8_to_binary(context_info, pvalue);
 
-                ((BinaryValue **) res->value)[i*2]   = bin_key;
-                ((BinaryValue **) res->value)[i*2+1] = bin_value;
+                if (!bin_key || !bin_value) {
+                    BinaryValueFree(bin_key);
+                    BinaryValueFree(bin_value);
+                    goto err;
+                }
+
+                res->hash_val[i * 2]     = bin_key;
+                res->hash_val[i * 2 + 1] = bin_value;
+                res->len++;
             }
-        } else {
-            // empty hash
-            res->len = 0;
-            res->value = NULL;
-        }
-        return res;
+        } // else empty hash
     }
 
-    Local<String> rstr = value->ToString();
+    else {
+        Local<String> rstr = value->ToString();
 
-    res->type = type_str_utf8;
+        res->type = type_str_utf8;
+        res->len = size_t(rstr->Utf8Length()); // in bytes
+        size_t capacity = res->len + 1;
+        res->str_val = static_cast<char *>(malloc(capacity));
+        if (!res->str_val) {
+            goto err;
+        }
 
-    int str_len = rstr->Utf8Length();
-    res->len = str_len;
-    str_len++;
-    res->value = (char *) malloc(str_len);
-    if (!res->value) {
-        free(res);
-        return NULL;
+        rstr->WriteUtf8(res->str_val);
     }
-    rstr->WriteUtf8((char *) res->value, str_len);
     return res;
+
+err:
+    BinaryValueFree(res);
+    return NULL;
 }
 
 
-void deallocate(void * data) {
+static void deallocate(void * data) {
     ContextInfo* context_info = (ContextInfo*)data;
     {
+        // XXX: what is the point of this?
```

Comment:
I think it acquires (and immediately loose) the lock on V8's context which make sure that nobody is working on the context at the very moment. Probably still open to some race conditions, but most likely avoid *most* horrors.

Reply:
I find it a bit bizarre that the lock doesn't extend further down. I'm leaving it as it is, though. I see ruby's mini_racer has similar code.

Reply:
Yeah, maybe `Reset` need a free lock? V8 codebase is a bit tedious to crawl through.